### ISSUE TITLE
Causal Convolutions in Tensorflow

### ISSUE THREAD

**Issue by joshim5** [tensorflow/tensorflow#23610]
**System information**
- TensorFlow version (you are using): 1.10
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
This is feature request to support **causal convolutions** in the Tensorflow layers API. Currently, causal convolutions require `tf.nn.conv2d(...)` with manual masking of weights. To improve speed, the data can be cropped/padded and then ordinary convolutions used.  

**Will this change the current api? How?**
Yes, `tf.layers.conv2d` will have a `causal` boolean flag.

**Who will benefit with this feature?**
Researchers and practitioners working on auto-regressive convolutional models.

**Any Other info.**
An analysis at OpenAI revealed that many open-source implementations of causal convolutions are slow, incorrect, and/or have uninterpretable code. This feature would help the ML community more easily develop architectures containing causal convolutions.

**Comment by mrry** [comment]
Adding this to the API would primarily impact Keras, so reassigning this to @fchollet.

**Comment by facaiy** [comment]
As far as I know, [ tf.keras.layers.Conv1D ](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv1D) supports causal padding. Does it work for you ?

**Comment by joshim5** [comment]
The issue in question relates to _2D Convolutions_, which are significantly harder to implement.

**Comment by omalleyt12** [comment]
I think generalizing the approach to causal padding used by `tf.keras.layers.Conv1D` to n-dimensional convolutions would be straight-forward enough if we assumed there was only one causal/"time" dimension. The 2-d case could be useful for generating audio spectrograms, the 3-d for video generation. We'd have to add an arg to the API for the user to specify which dimension is causal. Also, rather than adding a `causal` flag, Keras does causal convolutions with the `padding="causal"` arg.

The implementation of causal padding in `tf.keras.layers.Conv1d` uses the crop/pad strategy described in the feature request and could be generalized to this n-dim case. @joshim5 would you like to submit a PR for this? If not I can take a look

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @joshim5 ! Could you look at these similar issues for implementing Causal convolutions. [1,](https://stackoverflow.com/questions/60009083/timeseries-prediction-using-dilated-causal-convolutions-with-tensorflow-2-0) [2](https://colab.sandbox.google.com/github/tensorchiefs/dl_book/blob/master/chapter_02/nb_ch02_04.ipynb) .Thank you!

**Comment by RHimaJyothi** [comment]
Any update on this i.e., causality of conv2d? Or any other open source implementation for the causal Conv2D convolutions.
